controlpanel/api/auth0.py

```diff
@@ -0,0 +1,190 @@
+
+            items.extend(response[key])
+
+            if len(items) >= total:
+                break
+
+            if len(response[key]) < 1:
+                break
+
+            if "page" not in kwargs["params"]:
+                kwargs["params"]["page"] = 1
+
+            kwargs["params"]["page"] += 1
+
+        return items
+
+
+class AuthorizationAPI(APIClient):
+    base_url = settings.AUTH0["authorization_extension_url"]
+    audience = "urn:auth0-authz-api"
+
+    def get_group(self, group_name):
+        groups = [
+            group for group in self.get_all("groups") if group["name"] == group_name
+        ]
+
+        if len(groups) != 1:
+            return None
+
+        return groups[0]
+
+    def get_group_members(self, group_name):
+        group = self.get_group(group_name)
+        if group:
+            return self.get_all(
+                f'groups/{group["_id"]}/members', key="users", per_page=25
+            )
+
+    def add_group_members(self, group_name, emails, user_options={}):
+        group = self.get_group(group_name)
+        mgmt = ManagementAPI()
+        users = self.get_all("users", page=0, per_page=100)
+        user_lookup = {user["email"]: user for user in users if "email" in user}
+
+        def has_options(user):
+            for identity in user["identities"]:
+                if all(item in identity.items() for item in user_options.items()):
+                    return True
+
+        users_to_add = OrderedDict()
+
+        for email in emails:
+            user = user_lookup.get(email)
+
+            if user and has_options(user):
+                users_to_add[email] = user
+
+            else:
+                users_to_add[email] = mgmt.create_user(
+                    email=email, email_verified=True, **user_options
+                )
+
+        response = self.request(
+            "PATCH",
+            f'groups/{group["_id"]}/members',
+            json=[user["user_id"] for user in users_to_add.values()],
+        )
+
+        if "error" in response:
+            raise Auth0Error("add_group_members", response)
+
+    def delete_group_members(self, group_name, user_ids):
+        group = self.get_group(group_name)
+        if group is None:
+            return None
+
+        response = self.request(
+            "DELETE", f'groups/{group["_id"]}/members', json=user_ids
+        )
+
+        if "error" in response:
+            raise Auth0Error("delete_group_members", response)
+
+
+class ManagementAPI(APIClient):
+    base_url = f"https://{settings.AUTH0['domain']}/api/v2/"
+    audience = base_url
+
+    def create_user(self, email, email_verified=False, **kwargs):
+        response = self.request(
+            "POST",
+            "users",
+            json={"email": email, "email_verified": email_verified, **kwargs},
+        )
+
+        if "error" in response:
+            raise Auth0Error("create_user", response)
+
+        return response
```

Comment:
#### 1d. Refactor Auth0 client module

> Lines 176 - 190

...and for the Management API, providing a convenience method to create a user in Auth0.